kernel/kernel-api/src/main/java/io/delta/kernel/internal/table/ResolvedTableBuilderImpl.java

```diff
@@ -106,7 +111,41 @@ public ResolvedTableInternal build(Engine engine) {
 
   private void validateInputOnBuild() {
     checkArgument(ctx.versionOpt.orElse(0L) >= 0, "version must be >= 0");
-    // TODO: logData only ratified staged commits
-    // TODO: logData sorted and contiguous
+    validateProtocolAndMetadataOnlyIfVersionProvided();
+    validateProtocolRead();
+    validateLogDatasOnlyRatifiedCommits();
+    validateLogDatasSortedContiguous();
+  }
+
+  private void validateProtocolAndMetadataOnlyIfVersionProvided() {
+    checkArgument(
+        ctx.versionOpt.isPresent() || !ctx.protocolAndMetadataOpt.isPresent(),
+        "protocol and metadata can only be provided if a version is provided");
+  }
+
+  private void validateProtocolRead() {
+    ctx.protocolAndMetadataOpt.ifPresent(
+        x -> TableFeatures.validateKernelCanReadTheTable(x._1, ctx.unresolvedPath));
+  }
+
+  private void validateLogDatasOnlyRatifiedCommits() {
+    for (ParsedLogData logData : ctx.logDatas) {
+      checkArgument(
+          logData.type == ParsedLogType.RATIFIED_STAGED_COMMIT,
```

Comment:
do we (kernel) care about the differences between each kind of commit? are {ratified, normal, etc.} all ultimately treated the same or are there special considerations for each?